MAINT: a few more backports for 1.16.0

tylerjereddy opened the pull request:
A few more things to tie up before the final `1.16.0` release. Skipping CI for now, until we're ready to go.

Backports included (so far):

1. gh-23178
2. gh-23180
3. gh-23189
4. gh-23144
5. gh-23197
6. gh-23199

TODO:
- [x] handle the Python version check at build time per gh-23184 (Ralf or I will do this in `meson`)
- [x] bump the version of Python used for the array API testing job (should error out with current `3.10` once first point above is handled) on this branch (unless this happens to not be bumped on `main` either, in which case we could backport, but anyway fix that here either way) -- EDIT: need to backport gh-23144
- [x] check the status of CI here again
- [ ] maybe update release notes one more time to get final changes (i.e., scrape in gh-23199); if I do that, probably `[docs only]`

ev-br committed:
MAINT: array types: array-api-strict got stricter

ev-br committed:
CI: move array-api testing to python 3.12

rgommers committed:
BLD: drop rc1 from numpy build requirement, and bump to cibw 3.0.0

rgommers committed:
CI: remove workarounds for free-threaded wheels

Cython, Pythran and NumPy all have stable releases available now.

[ci skip] [skip ci]

crusaderky committed:
CI: skip JAX 0.6.2

rgommers committed:
BLD: implement build-time version check for minimum Python version

Closes gh-23184

tylerjereddy commented:
Thanks Evgeni, I don't think you need to open a PR, I've added it to the TODO list above. I'm a bit tied up with grant reviews today and tomorrow, but sounds like the path forward is there.

ilayn committed:
DEP: optimize: Add deprecation warnings to L-BFGS-B `disp`, `iprint`

tylerjereddy committed:
MAINT, TST: NumPy stride deprecation shim

* This patch allows the SciPy testsuite to pass against
NumPy `main` branch, after https://github.com/numpy/numpy/pull/28925 was
merged in the last day or so.

tylerjereddy committed:
CI, MAINT: PR 23188 revisions

* Remove the upper bound on the `array-api-strict` version
used in CI testing, now that some backports have been added in
to fix related issues.

tylerjereddy committed:
DOC: PR 23188 revisions

* Update the SciPy `1.16.0` release notes following
additional backport activity.

tylerjereddy committed:
Revert "CI, MAINT: PR 23188 revisions"

This reverts commit 5a1775fe709d72bf91a345c429121d406a028573.

tylerjereddy commented:
Backporting gh-23199 also caused an issue that didn't show up on `main`. I need to move on for now, I'll try to revisit this fresh tomorrow.

tylerjereddy commented:
Do you see anything in `git diff main -- .github/workflows/array_api.yml` that could explain that?

Coming from another "included" file perhaps? Must be something obvious I'm missing on that one.

lucascolley commented:
I am completely puzzled after staring at the files for a few minutes.

ev-br commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1295,7 +1295,7 @@ def test_odd_strides(self, xp):
         dummy = xp.arange(10, dtype=xp.float64)
         a = dummy[5:6]
-        a.strides = 16
+        a = xp.lib.stride_tricks.as_strided(a, strides=(16,))
```
```suggestion
        a = np.lib.stride_tricks.as_strided(a, strides=(16,))
```

because otherwise (https://github.com/scipy/scipy/actions/runs/15788333110/job/44509424300?pr=23188), grumble grumble

```
>       a = xp.lib.stride_tricks.as_strided(a, strides=(16,))
            ^^^^^^
E       AttributeError: module 'scipy._lib.array_api_compat.numpy' has no attribute 'lib'
```

lucascolley replied:
gh-23199 now merged as so

tylerjereddy committed:
TST, MAINT: PR 23188 revisions

* Apply the reviewer-suggested shim to `TestMedFilt::test_odd_strides`,
so that this test may pass with both older and newer versions of NumPy.

[ci skip] [skip ci]

tylerjereddy commented:
I applied the suggested change from https://github.com/scipy/scipy/pull/23188#discussion_r2159983946, so that the testsuite may pass with older and newer versions of NumPy.

For now though, I've skipped the CI again because the shenanigans from https://github.com/scipy/scipy/pull/23188#issuecomment-2993011648 are reproducing with manual CI job restarts and need a closer look.

tylerjereddy committed:
CI: PR 23188 revisions

* try disabling the Python version cache in the array API
CI job to deal with https://github.com/scipy/scipy/pull/23188#issuecomment-2993011648

[skip circle]

tylerjereddy commented:
oh yuck (see below), that must be the problem... I'll try to patch it out and probably forward port to `main` too if that removal fixes it--I can't think of a reason we'd want this anymore

```yaml
    strategy:
      matrix:
        python-version: ['3.12']
        maintenance-branch:
          - ${{ contains(github.ref, 'maintenance/') || contains(github.base_ref, 'maintenance/') }}
        exclude:
          - maintenance-branch: true
```

tylerjereddy committed:
Revert "CI: PR 23188 revisions"

This reverts commit d41aa3a4361b6723c2d3112180b392bf7b8ef6a7.

tylerjereddy committed:
CI: PR 23188 revisions [wheel build]

* Stop excluding the maintenance branch from Python version
setting in the `matrix` directive of our array API CI test
job, to deal with:
https://github.com/scipy/scipy/pull/23188#issuecomment-2993011648

[wheel build]

tylerjereddy commented:
The full diff looks exactly as I'd expect it to now. I took a bit of a risk and started a full CI run, including wheels, with the latest fixes.

If that turns out well, the only other thing I might want to do here is update the release notes one more time, although the changes from doing so are almost not worth it (basically just pulling in gh-23199 to the list since it was recently merged after its backport happened here).

tylerjereddy commented:
I think the single prerelease job failure may be "ok," as long as all the wheel builds (and everything else) are green.

NumPy has been picking up on the merging of bigger changes after their recent release I think, so the new warning could be related to that for example. Handling `main` branch NumPy changes on this release branch is lower prio.

tylerjereddy commented:
Merging based on analysis above, we need to prioritize getting the release out over minor non-blocking matters at this point I think. The 1 pre-release test failure seems safe to ignore for the moment.
